### ISSUE TITLE
Is there a way to set an allow_domain function per server?

### ISSUE THREAD

**Issue by brendon** [GUI/lua-resty-auto-ssl#73]
I was just writing up a function to query a database for wether a particular domain is allowed to have an SSL cert issued and realised that I have a couple of other `server` definitions on this server with specific domains in addition to my catch-all server and that these names would have to be checked also.

This won't really be a problem as I can just add a regex check against a list before I go to query the database, but I wondered if it's possible to override the `allow_domain` function per server? I'm new to lua so that's probably where my understanding is falling down. It looks like a function on an object: `auto_ssl`, so I assume there is only room for one function?

**Comment by luto** [comment]
Currently `allow_domain` can only be specified once for the entire nginx instance. But it's certainly possible to change that. What's the motivation behind your request? Are you trying to keep the domain configuration separate from another to preserve a plug-and-play-ish configuration?

**Comment by brendon** [comment]
Yes that's pretty much correct. On my server I have a config file for `auto_ssl` in the `config.d` directory, then for each application I have a file in `sites_enabled`. Each of these site files has:

```
# Dynamic handler for issuing or returning certs for SNI domains.
  ssl_certificate_by_lua_block {
    auto_ssl:ssl_certificate()
  }
  ssl_certificate /etc/ssl/resty-auto-ssl-fallback.crt;
  ssl_certificate_key /etc/ssl/resty-auto-ssl-fallback.key;
```

and it seemed to make more sense to be able to also configure the allowed domains per `server`. In my case for most `server`'s they're pre-known, but for one I need to use mysql to query for wether I care about the domain or not.

**Comment by luto** [comment]
So, I played around a bit. The main problem in actually implementing this is that there is no way to hook into the initialization process of each `server{}` block. I also could not find an API to determine which server block actually invoked the currently running `ssl_certificate_by_lua`. So this has to be done within the request cycle.

According to [the fancy openresty request flowchart](https://cloud.githubusercontent.com/assets/2137369/15272097/77d1c09e-1a37-11e6-97ef-d9767035fc3e.png) there is no earlier hook than `ssl_certificate_by_lua`. Thus, the most straight forward way to achieve our goal here seems to be something like this:
```lua
server {
  listen 443 ssl;

  ssl_certificate_by_lua_block {
    auto_ssl:ssl_certificate(function(domain)
      return true  -- insert magic here
    end)
  }
```

Does look too bad. I just wonder what kind of performance hit you'd get for newly defining the function within each request. Implementing this could work in a backwards-compatible way: Still allow a global `allow_domain` function to be set, but make the function passed to `auto_ssl:ssl_certificate()` overwrite it.

The actual work also doesn't seem to be too hard. Add the new parameter to [`auto_ssl:ssl_certificate()`](https://github.com/GUI/lua-resty-auto-ssl/blob/3412d18a99937b258fc4218c4fefc2348d5a7c84/lib/resty/auto-ssl.lua#L72-L75), as well as [`resty.auto-ssl.ssl_certificate`](https://github.com/GUI/lua-resty-auto-ssl/blob/3412d18a99937b258fc4218c4fefc2348d5a7c84/lib/resty/auto-ssl/ssl_certificate.lua#L262-L267) and [`do_ssl`](https://github.com/GUI/lua-resty-auto-ssl/blob/3412d18a99937b258fc4218c4fefc2348d5a7c84/lib/resty/auto-ssl/ssl_certificate.lua#L228). Then determine which `allow_domain` function to call - the global or the server one - [within `do_ssl`](https://github.com/GUI/lua-resty-auto-ssl/blob/3412d18a99937b258fc4218c4fefc2348d5a7c84/lib/resty/auto-ssl/ssl_certificate.lua#L237-L242) and, well, call it.

@GUI what's your take on this?

**Comment by brendon** [comment]
Thanks @luto, yes I guess the only real concern is the cost of creating that function over and over. It's a shame about the hooks though. Do you think there's a case for raising a ticket with `lua-nginx` about adding another hook?

**Comment by luto** [comment]
```lua
function add(a, b) return a + b end
for i=0,10000000000,1 do add(2, 5) end
```

Takes 50 seconds to run :arrow_right: 50 nanoseconds for each call.

```lua
for i=0,1000000000,1
do
  (function (a, b)
    return a + b
  end)(2, 5)
end
```

Takes 13s to run :arrow_right: 1.2ns for each call.

Both times were measured using `time(1)` on an Intel E5-2630L processor.

The _difference_ between those two is significant. However, 50 nanoseconds is nothing. Not even scratching on 1ms here. Seeing those numbers, I do not have any concerns about creating that function over and over.

**Comment by brendon** [comment]
Indeed yes it seems to be minor :) It all sounds good to me.

**Comment by luto** [comment]
@brendon are you up for preparing a pull request? :) In case you need anything or the some documentation is unclear or missing, just ping me!

**Comment by brendon** [comment]
Haha! I've never written lua before (except for the `allow_domain` function I wrote) but I can have a crack. I'm sure it won't be too hard given the general gist of the new code is already there :)

**Comment by brendon** [comment]
Unfortunately I can't run docker on My Late 2008 Macbook Pro :( I suppose I can still give things a crack, but I won't be able to test anything. What do you think?

The wheels are starting to fall off this laptop. Such a shame since it still runs really well :)

**Comment by GUI** [comment]
I think all this makes sense, and it does seem like having the option to define this within the server context could be handy.

One thing to note is that `ssl_certificate()` already accepts a table of options (see the examples under the `request_domain` docs on the README). So it might make sense to make this callback another item in that options table, or a separate argument.

Regarding tests, you don't necessarily need Docker to run them (that's just what I had documented, hoping it was the easiest). If you have a local development environment where OpenResty is running, I think you should be able to run tests on your Mac via `make test` (but if you're not able to get them running, I'm always happy to take a look at any pull requests).

**Comment by brendon** [comment]
I actually managed to get the docker thing working with docker-toolbox instead (for older Mac's). I'll hopefully be able to work on this next week.

**Comment by brendon** [comment]
Just kicked off the tests but I get the following:

```
cd /test/run/tmp && rm -rf luarocks*
cd /test/run/tmp && curl -L -O http://luarocks.org/releases/luarocks-2.4.2.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   167  100   167    0     0    475      0 --:--:-- --:--:-- --:--:--   528
100  135k  100  135k    0     0   150k      0 --:--:-- --:--:-- --:--:--  150k
cd /test/run/tmp && tar -xf luarocks-2.4.2.tar.gz
cd /test/run/tmp/luarocks-2.4.2 && ./configure \
	--prefix=/test/run/build/luajit \
	--with-lua=/test/run/build/luajit \
	--with-lua-include=/test/run/build/luajit/include/luajit-2.1 \
	--lua-suffix=jit-2.1.0-beta2
Illegal instruction
Lua version detected: 
Lua interpreter found: /test/run/build/luajit/bin/luajit-2.1.0-beta2...
Checking Lua includes... lua.h found in /test/run/build/luajit/include/luajit-2.1/lua.h
Checking if Lua header version matches that of the interpreter... no
lua.h version mismatch (interpreter: ; lua.h: 5.1).
You may want to use the flag --with-lua or --with-lua-include. See --help.

configure failed.

make: *** [Makefile:186: /test/run/tmp/stamp-luarocks-2.4.2] Error 1
```

That just on `master` for now. I've not made any changes. I've just run `docker-compose run app make test`.

Do I need to do anything special with docker before running that command?

**Comment by luto** [comment]
No, the docker command should just.. work :-/ @GUI any idea here?

**Comment by brendon** [comment]
Just pinging this quickly. I've really not had any time to work on it unfortunately.
